From dda8157564bf82593f85fea75b5eedeb12b1c101 Mon Sep 17 00:00:00 2001
From: Zhiyi Zhang <zzhang@codeweavers.com>
Date: Tue, 20 Feb 2024 11:04:13 +0800
Subject: [PATCH] dsound: Reject WAVEFORMATEX formats with more than two
 channels.

Formats with more than two channels require WAVEFORMATEXTENSIBLE according to tests.

Fix Viking: Battle for Asgard (211160) audio cracking in its intro video.

(cherry picked from commit 7c7b2e8e7e8c0be1a756e6d7c705b68017459e26)

CW-Bug-Id: #19927
---
 dlls/dsound/dsound.c       | 3 +++
 dlls/dsound/primary.c      | 3 +++
 dlls/dsound/tests/dsound.c | 2 --
 3 files changed, 6 insertions(+), 2 deletions(-)

diff --git a/dlls/dsound/dsound.c b/dlls/dsound/dsound.c
index b1c0c76c367..247d1edcf6b 100644
--- a/dlls/dsound/dsound.c
+++ b/dlls/dsound/dsound.c
@@ -489,6 +489,9 @@ static HRESULT DirectSoundDevice_CreateSoundBuffer(
             return DSERR_INVALIDPARAM;
         }
 
+        if (dsbd->lpwfxFormat->nChannels > 2 && dsbd->lpwfxFormat->wFormatTag != WAVE_FORMAT_EXTENSIBLE)
+            return DSERR_INVALIDPARAM;
+
         if (dsbd->lpwfxFormat->wFormatTag == WAVE_FORMAT_EXTENSIBLE)
         {
             WAVEFORMATEXTENSIBLE *pwfxe = (WAVEFORMATEXTENSIBLE*)dsbd->lpwfxFormat;
diff --git a/dlls/dsound/primary.c b/dlls/dsound/primary.c
index 8cfcd1d8a28..22cf475a674 100644
--- a/dlls/dsound/primary.c
+++ b/dlls/dsound/primary.c
@@ -538,6 +538,9 @@ HRESULT primarybuffer_SetFormat(DirectSoundDevice *device, LPCWAVEFORMATEX passe
 			return DSERR_INVALIDPARAM;
 	}
 
+        if (passed_fmt->nChannels > 2 && passed_fmt->wFormatTag != WAVE_FORMAT_EXTENSIBLE)
+            return DSERR_ALLOCATED;
+
 	/* **** */
 	AcquireSRWLockExclusive(&device->buffer_list_lock);
 	EnterCriticalSection(&(device->mixlock));
